integration_tests/src/main/python/data_gen.py

```diff
@@ -27,12 +27,44 @@
 from conftest import skip_unless_precommit_tests
 import time
 import os
+from functools import lru_cache
+import hashlib
 
 # set time zone to UTC for timestamp test cases to avoid `datetime` out-of-range error:
 # refer to: https://github.com/NVIDIA/spark-rapids/issues/7535
 os.environ['TZ'] = 'UTC'
 time.tzset()
 
+# hash object with nested data types for caching
+def hash_object(obj):
+    h = hashlib.blake2b(digest_size=10)
+    if isinstance(obj, (int, float, str, bool, type(None))):
+        # convert to bytes-like object
+        h.update(str(obj).encode())
+    elif isinstance(obj, (list, tuple)):
+        for item in obj:
+            h.update(hash_object(item).encode())
+    elif isinstance(obj, dict):
+        keys = sorted(obj.keys())  
+        for k in keys:     
+            h.update(hash_object(k).encode()) 
+            h.update(hash_object(obj[k]).encode())
+    elif isinstance(obj, DataGen):
+        h.update(obj.__class__.__name__.encode())
+        attrs = vars(obj)
+        keys = sorted(attrs.keys())
+        for k in keys:
+            if str(k) == '_gen_func':
+                # bypass internal functions
+                continue
+            h.update(hash_object(k).encode())
+            h.update(hash_object(attrs[k]).encode())
+    elif callable(obj):
+        h.update(obj.__code__.co_code)
+    else:
+        h.update(str(hash(obj)).encode())
```

Comment:
is this where `spark` SparkSession  is handled?

looks like hash of SparkSession does not reflect the current runtime config. If we can reason it's ok in use case please add some comment. Otherwise, we might need an issue to handle configs affecting createDataFrame

Reply:
`@lru_cache` will handle the lookup of SparkSession in hash with it's `__eq__` method, and the current runtime config will not affect the equality of it. The `hash_object` will only used to check the equality of `DataGen`s.

Reply:
> the current runtime config will not affect the equality of it

and thus we need to be able to reason why or when is our use of createDataFrame produces identical results irrespective of the runtime config. There is at least one conditional already at the Python level https://github.com/apache/spark/blob/c3b62708cd6371da08943e572a9bc0a45c1dcced/python/pyspark/sql/session.py#L1332, and maybe more at the underlying Scala as well

Reply:
I did some tests, shows that the memory address (which python check for equality) of sparksession will reflect the current runtime configs. And since there is only one sparksession at the same time, we can use the memory address to cache spark with the same configs. I added a comment in `gen_df`.

Reply:
When spark.conf is mutated its address stays the same as well. But the createDataFrame result may differ. 

Here is a notebook with a contrived example demonstrating the issue https://github.com/gerashegalov/rapids-shell/blob/master/src/jupyter/create-dataframe-conf.ipynb

It would be safer to add spark._jconf.getAllConfs().toString() as an input to the hash key. Or have some logic checking that rejecting such confs at gen_df time.

Reply:
I did more tests, you are right, thank you! Different configs do not affect the memory address, I was misled by the results of the parallel tests. I added `spark._jconf.getAllConfs().toString()` as cache key now. The results show that the cache hits decreased after this change, and I'm curious why all the IT passed when not considering this, will continue to find out why...

Reply:
The change is conservatively pessimistic including the diffs in the configs that will not affect createDataFrame results.  I premised the thread on the need to reason which configs might be problematic.

> The results show that the cache hits decreased

One of the reasons is that with the pessimistic approach gen_df_help result is not cached between GPU and CPU since at least spark.rapids.sql.enabled values are different. It should be safe to ignore `spark.rapids.sql.enabled`  and `spark.rapids.sql.test.*` https://github.com/NVIDIA/spark-rapids/blob/291a389cc85ae8dbc609c50559de951a38b131da/integration_tests/src/main/python/spark_session.py#L112-L133

> why all the IT passed when not considering this

possible reasons
- current config variations don't cause issues 
- maybe tests for such edge cases just don't exist or scenarios not really valid
- we often compare results after converting them to Python-land arrays, it may cause discrepancies. In the contrived example `spark.sql('SELECT ts_ntz FROM df1').collect() == spark.sql('SELECT ts_ltz FROM df2').collect()` is True regardless of the conf while in SQL it will depend on the current timeZone.

Reply:
Thanks, learned a lot. 

`SparkSession` object has no attribute `_jconf` in 3.1.1 so CI failed. So I think maybe we can cache the data list instead of caching dataframe for now to keep fine speedup without causing problems about sparkconf. 

I updated the code, please have a look. We can file a follow up issue about caching dataframe properly (considering the sparkconf issue) if you think caching data is ok for now.

Result for `TEST_PARALLEL=1 ./run_pyspark_from_build.sh -k 'cmp'` after this change: `430 passed, 19545 deselected, 1 warning in 199.97s (0:03:19)`

Reply:
+1 for a follow-up issue. I am fine going back to df caching in the meantime if it provides significant savings. but this LGTM too

Reply:
Filed follow-up issue #8524

Reply:
I don't understand how the configs are going to impact the data gen.  The data is not generated by Spark. The caching happens before the data is even handed to spark to produce a data frame.  We don't really support time zones yet, but when we do I would assume that it would be one of the parameters to the data gen.